File: LeetCode/LRU_cache.py
#!/usr/bin/python
class ListNode:
  def __init__(self, key, value):
    self.key = key
    self.value = value
    self.next = None
    self.pre = None

class LRUCache:

    # @param capacity, an integer
    def __init__(self, capacity):
      self.cap = capacity
      self.n = 0
      self.head = ListNode(-1, -1)
      self.tail = self.head

    # @return an integer
    def get(self, key):
      node = self.head.next
      while node != None and node.key != key:
        node = node.next
      if node == None:
        return -1
      else:
        if node == self.head.next: return node.value
        if node == self.tail:
          node.pre.next = None
          self.tail = node.pre
          node.pre = self.head
          node.next = self.head.next
          self.head.next.pre = node
          self.head.next = node
        else:
          node.pre.next = node.next
          node.next.pre = node.pre
          node.next = self.head.next
          self.head.next.pre = node
          node.pre = self.head
          self.head.next = node
      return node.value

    # @param key, an integer
    # @param value, an integer
    # @return nothing
    def set(self, key, value):
      if self.get(key) == -1:
        self.n += 1
        node = ListNode(key, value)
        if self.n == 1:
          self.head.next = node
          node.pre = self.head
          self.tail = node
        else:
          self.head.next.pre = node
          node.next = self.head.next
          self.head.next = node
          node.pre = self.head
        if self.n > self.cap:
          self.n -= 1
          self.tail.pre.next = None
          self.tail = self.tail.pre
      else:
        node = self.head.next
        while node.key != key:
          node = node.next
        node.value = value

ss = LRUCache(2)
ss.set(2, 1)
print ss.get(2)

File: LeetCode/LRU_cache.cpp
/*
ID: ck89119
PROG:
LANG: C++
*/
#include <vector>
#include <list>
#include <map>
#include <set>
#include <deque>
#include <queue>
#include <stack>
#include <string>
#include <bitset>
#include <algorithm>
#include <functional>
#include <numeric>
#include <utility>
#include <sstream>
#include <fstream>
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cmath>
#include <cstring>
#include <cstdlib>
using namespace std;
#define dump(x) cerr <<  __LINE__ << " : "<< #x << "  =  " << (x) <<endl
#define FR(i,a,b) for(int i=(a);i<(b);++i)//[a,b)
#define CLR(a,x) memset(a,x,sizeof(a)) //0:0, -1:-1, INF:0x3f, -INF:0x80
#define MP(A,B) make_pair(A,B)
#define PB(A) push_back(A)
#define PAU system("pause")
#define lson l, m, rt << 1
#define rson m + 1, r, rt << 1 | 1
#define lowbit(n) (n^(n-1))&n
#define fin freopen(".in","r",stdin)
#define fout freopen(".out","w",stdout)
//ifstream fin("");
//ofstream fout("");
#define LL long long

template <class T> void out(T A[],int n){for (int i=0;i<n;i++) cout<<A[i]<<" ";cout<<endl;}
template <class T> void out (vector<T> A,int n=-1){if(n==-1) n=A.size();for (int i=0;i<n;i++) cout<<A[i]<<" ";cout<<endl;}
const int MAXN=99999;
const int MAXM=9999999;
const int INF=0x3f3f3f3f;
const double PI=acos(-1.0);
const double EPS=1e-11;

struct DoubleLinkedNode {
  int key, value;
  struct DoubleLinkedNode *next, *pre;
  DoubleLinkedNode(int _key = -1, int _value = -1) {
    key = _key;
    value = _value;
    next = pre = NULL;
  }
};

class LRUCache{
public:
    DoubleLinkedNode *head, *tail;
    int cap, size;
    map<int, struct DoubleLinkedNode*> h;

    LRUCache(int capacity) {
      cap = capacity;
      size = 0;
      head = new DoubleLinkedNode(-INF, -1);
      tail = new DoubleLinkedNode(INF, -1);
      head->next = tail;
      tail->pre = head;
      //h[-INF] = head;
      //h[INF] = tail;
    }
    
    DoubleLinkedNode *removeNode(int key) {
      DoubleLinkedNode *tmp = h[key];
      DoubleLinkedNode *pre = tmp->pre;
      DoubleLinkedNode *next = tmp->next;
      pre->next = next;
      next->pre = pre;
      h.erase(key);
      return tmp;
    }
    
    int pushNode(DoubleLinkedNode *a) {
      tail->pre->next = a;
      a->pre = tail->pre;
      tail->pre = a;
      a->next = tail;
      h[a->key] = a;
      return 0;
    }
    
    DoubleLinkedNode *popNode() {
      if (head->next == tail) return NULL;
      DoubleLinkedNode *a = head->next;
      head->next = a->next;
      a->next->pre = head;
      h.erase(a->key);
      return a;
    }

    int get(int key) {
      if (h.count(key) == 0) return -1;
      else {
        DoubleLinkedNode *tmp;
        tmp = removeNode(key);
        pushNode(tmp);
        return tmp->value;
      }
    }
    
    void set(int key, int value) {
      if (h.count(key) == 0) {
        if (size == cap) popNode();
        DoubleLinkedNode *a = new DoubleLinkedNode(key, value);
        pushNode(a);
      } else {
        DoubleLinkedNode *a;
        a = removeNode(key);
        a->value = value;
        pushNode(a);
      } 
    }
};
